server/src/main/java/io/crate/statistics/ReservoirSampler.java

```diff
@@ -89,20 +93,59 @@ public final class ReservoirSampler {
     private CircuitBreakerService circuitBreakerService;
     private final IndicesService indicesService;
 
+    private volatile ByteSizeValue readLimit;
+    private volatile ByteSizeValue updatedReadLimit;
+    private volatile boolean analyzing;
+
+    private final RateLimiter rateLimiter;
+    private long bytesSinceLastRateLimit = 0;
+
     @Inject
     public ReservoirSampler(ClusterService clusterService,
                             NodeContext nodeCtx,
                             Schemas schemas,
                             CircuitBreakerService circuitBreakerService,
-                            IndicesService indicesService) {
+                            IndicesService indicesService,
+                            Settings settings) {
         this.clusterService = clusterService;
         this.nodeCtx = nodeCtx;
         this.schemas = schemas;
         this.circuitBreakerService = circuitBreakerService;
         this.indicesService = indicesService;
+
+        this.readLimit = STATS_SERVICE_THROTTLING_SETTING.get(settings);
+        this.rateLimiter = new RateLimiter.SimpleRateLimiter(readLimit.getMbFrac());
+        this.updatedReadLimit = readLimit;
+        this.analyzing = false;
+
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(
+            STATS_SERVICE_THROTTLING_SETTING, this::updateReadLimit);
+    }
+
+    private void updateReadLimit(ByteSizeValue newReadLimit) {
+        boolean isAnalyzing = analyzing; // One volatile read.
+        if (isAnalyzing == false) {
```

Comment:
why can't we adjust the RateLimiter directly? Why do we have to wait until a running analyze finishes?

Reply:
replied in https://github.com/crate/crate/pull/13046#discussion_r976408756